src/main/java/org/codehaus/mojo/servicedocgen/ServiceDocGenReport.java

```diff
@@ -62,259 +39,19 @@
-            }
-            else if ( file.getName().endsWith( ".java" ) )
-            {
-                try
-                {
-                    JavaSource source = builder.addSource( file );
-                    for ( JavaClass type : source.getClasses() )
-                    {
-                        boolean isService = isServiceClass( type );
-                        if ( isService )
-                        {
-                            serviceClasses.add( type );
-                        }
-                    }
-                }
-                catch ( Exception e )
-                {
-                    getLog().debug( "Error parsing file: " + file, e );
-                }
-            }
-        }
-
-    }
-
-    private boolean isServiceClass( JavaClass type )
-    {
-        if ( this.classnamePattern.matcher( type.getName() ).matches() )
-        {
-            getLog().debug( "Class matches: " + type.getName() );
-            for ( JavaAnnotation annotation : type.getAnnotations() )
-            {
-                if ( Path.class.getName().equals( annotation.getType().getFullyQualifiedName() ) )
-                {
-                    getLog().info( "Found service: " + type.getName() );
-                    return true;
-                }
-            }
-        }
-        return false;
-    }
-
-    private ClassLoader getProjectClassloader()
-        throws MojoExecutionException
-    {
-        if ( this.projectClassloader == null )
-        {
-            this.projectClassloader = new URLClassLoader( buildClasspathUrls(), this.getClass().getClassLoader() );
-        }
-        return this.projectClassloader;
-    }
-
-    private URL[] buildClasspathUrls()
-        throws MojoExecutionException
-    {
-        List<URL> urls = new ArrayList<URL>( this.runtimeClasspathElements.size() );
-        for ( String element : this.runtimeClasspathElements )
-        {
-            try
-            {
-                URL url = new File( element ).toURI().toURL();
-                urls.add( url );
-                getLog().debug( "Adding to classloader: " + url.toString() );
-            }
-            catch ( MalformedURLException e )
-            {
-                throw new MojoExecutionException( "Unable to access project dependency: " + element, e );
-            }
-        }
-        return urls.toArray( new URL[urls.size()] );
-    }
+@Mojo(name = "generate", defaultPhase = LifecyclePhase.PREPARE_PACKAGE, requiresProject = true, requiresDirectInvocation = false, executionStrategy = "once-per-session", requiresDependencyCollection = ResolutionScope.COMPILE_PLUS_RUNTIME)
+public class ServiceDocGenReport extends ReportGen {
+  @Parameter(defaultValue = "Service-Documentation.html.vm")
+  private String templateName;
+
+  /**
+   * The constructor.
+   *
+   * @param templateName
+   */
+  public ServiceDocGenReport(String templateName) {
+
+    super(templateName);
+  }
```

Comment:
It is a little confusing that both the new super-class `ReportGen` as well as both its subclasses have the same `field` and all with `@Parameter` annotation (what would actually indicate maven to have 2 parameters with the same name and then it is unclear if the parameter of the sub-class or the super-class is set - in the worst case this will change with a newer maven version and break the plugin code).
Did you test this after all?
AFAIK a Maven MOJO *must* have a non-arg constructor.
So I see two options here:
1. Remove the `templateName` from the super-class (both field and constructor arg) and instead add an abstract and protected method `getTemplateName()` to use instead of the field. Then the two super-classes can both add the `@Parameter` and provide it via the getter.
2. Remove the `defaultValue` from the super-class and remove the field `templateName` from both sub-classes. Pass the according default value as string literal directly from the non-arg-constructors to the String-arg constructor of the super-class to init the different defaults.

I think option 1. is better as Maven generates documentation for every report and this approach 1. will still have the defaultValue in the documentation. Otherwise users will see a required parameter without defaults and might wonder what they have to configure here.